### PR TITLE
Add Icons to Completions

### LINKED ISSUE #2327: Add icons to completions
Add icons to the the completions based on type (a la [pycharm](https://www.jetbrains.com/pycharm/)).

We would return a list of (completion, type) tuples for a completion, and display the icon in the completion list.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #2327.

screen shot 2015-04-17 at 9 56 02 pm

### CODE REVIEW

**Comment by blink1073** [comment]
Note that neither Jedi nor Rope return `method` as a type, only `function`.

**Comment by goanpeca** [comment]
That means that methods will be listed as functions? or?

What about the privacy of the elements, anything to also have that?

**Comment by blink1073** [comment]
Methods are listed as functions.  The privacy can be inferred from the name, but the type cannot. I say the less icons the better.

**Comment by goanpeca** [comment]
Seems fine for now. It would be great to only display matching options instead of the full list

**Comment by blink1073** [comment]
That is for another PR.

**Comment by goanpeca** [comment]
Indeed

**Comment by ccordoba12** [comment]
Cool, cool, really cool!! I've wanted to have this for years!!

**Comment by ccordoba12** [comment]
A note about the new icons here: please add them to our LICENSE file :-)

File: spyderlib/utils/introspection/jedi_plugin.py [hunk 1]
```diff
@@ -54,10 +54,9 @@ def load_plugin(self):
         self._warmup_thread.start()
 
     def get_completions(self, info):
-        """Return a list of completion strings"""
+        """Return a list of (completion, type) tuples"""
         completions = self.get_jedi_object('completions', info)
-        debug_print(str(completions)[:100])
```
Comment (ccordoba12) [inline spyderlib/utils/introspection/jedi_plugin.py]: Why did you remove this `debug_print` here? Just curious, I don't how important that is.

File: spyderlib/utils/introspection/jedi_plugin.py [hunk 2]
```diff
@@ -267,7 +266,7 @@ def preload(self):
     source_code = "import n"
     completions = p.get_completions(CodeInfo('completions', source_code,
         len(source_code)))
-    assert 'numpy' in completions
+    assert ('numpy', 'module') in completions
```
Comment (ccordoba12) [inline spyderlib/utils/introspection/jedi_plugin.py]: Do we have an icon for modules?

File: spyderlib/utils/introspection/rope_plugin.py [hunk 3]
```diff
@@ -93,10 +93,11 @@ def get_completions(self, info):
             proposals = rope.contrib.codeassist.sorted_proposals(proposals)
             if DEBUG_EDITOR:
                 log_dt(LOG_FILENAME, "code_assist/sorted_proposals", t0)
-            return [proposal.name for proposal in proposals]
+            return [(proposal.name, proposal.type) for proposal in proposals]
         except Exception as _error:  #analysis:ignore
             if DEBUG_EDITOR:
                 log_last_error(LOG_FILENAME, "get_completion_list")
+        return []
```
Comment (ccordoba12) [inline spyderlib/utils/introspection/rope_plugin.py]: Why this addition here?

File: spyderlib/widgets/sourcecode/base.py [hunk 4]
```diff
@@ -61,15 +63,26 @@ def setup_appearance(self, size, font):
...
         if len(completion_list) == 1 and not automatic:
             self.textedit.insert_completion(completion_list[0])
             return
-        
+
         self.completion_list = completion_list
         self.clear()
-        self.addItems(completion_list)
+
+        lut = dict(instance='attribute.png', method='method.png',
```
Comment (ccordoba12) [inline spyderlib/widgets/sourcecode/base.py]: What this `lut` variable means? I mean, could you find a better name for it? Or at least a comment above about why you decided to name it like that?

File: spyderlib/widgets/sourcecode/base.py [hunk 5]
```diff
@@ -61,15 +63,26 @@ def setup_appearance(self, size, font):
...
+                   statement='attribute.png', function='function.png',
+                   attribute='attribute.png')
+        lut['class'] = 'class.png'
```
Comment (ccordoba12) [inline spyderlib/widgets/sourcecode/base.py]: Why this line is out of the previous `dict`?

File: spyderlib/widgets/sourcecode/base.py [hunk 6]
```diff
@@ -882,15 +895,17 @@ def hide_completion_widget(self):
...
         # Sorting completion list (entries starting with underscore are 
         # put at the end of the list):
-        underscore = set([comp for comp in completions
+        underscore = set([(comp, t) for (comp, t) in completions
                           if comp.startswith('_')])
-        completions = sorted(set(completions)-underscore, key=str_lower)+\
-                      sorted(underscore, key=str_lower)
+
+        completions = sorted(set(completions)-underscore, key=lambda x: str_lower(x[0]))+\
+                      sorted(underscore, key=lambda x: str_lower(x[0]))
```
Comment (ccordoba12) [inline spyderlib/widgets/sourcecode/base.py]: Could you break this line in two or more assignments? It's not easy to understand what it does :-)

**Comment by ccordoba12** [comment]
Ok, first review done.

File: spyderlib/utils/introspection/rope_plugin.py (hunk 3 above)
Comment (blink1073) [inline spyderlib/utils/introspection/rope_plugin.py]: To make sure we always return a list (I wasn't sure what else might break if I didn't).

File: spyderlib/utils/introspection/jedi_plugin.py (hunk 2 above)
Comment (blink1073) [inline spyderlib/utils/introspection/jedi_plugin.py]: We do not.  We have an "M" icon for method, but if we used that it would be the same one as the one used in the outline explorer.  I could make an "M" with a different color and use it for module.

**Comment by blink1073** [comment]
@ccordoba12, comments addressed.

**Comment by goanpeca** [comment]
Merge merge merge merge :+1:

**Comment by goanpeca** [comment]
I can make a new set of icons (pngs.) for this after merged, I am ok doing that (now I have some xp), so @blink1073 can focus on functionality and not on looks, ok?

My idea of the privacy is not to include more icons (indeed can become confusing), the point is that if we can for now return that info we could eventually use it to make some things... like never display super private methods in completion outside the module they are defined (as an option in the global preferences), or only display private methods if you are inside the actual class defining them. Also maybe Python special methods should be always on top?

**Comment by blink1073** [comment]
I don't see a need to get any more complicated than what we have.  The underscore methods are moved to the bottom by default because you are more often looking for other attributes.  Also, Python is built on the "we're all consenting adults" mantra, which arbitrarily hiding private methods would go against that.

**Comment by goanpeca** [comment]
ok

**Comment by goanpeca** [comment]
On the icons I meant for the current ones, which look pixelated... I do not intend to make new icons, but make the existing ones more uniform.

**Comment by SylvainCorlay** [comment]
@goanpeca or just make bitmap versions of the new qtawesome ones, once I made them pixel perfect.

File: spyderlib/utils/introspection/jedi_plugin.py (hunk 2 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/jedi_plugin.py]: I think that would be confuse. What about using a (P) icon for package?

File: spyderlib/utils/introspection/rope_plugin.py (hunk 3 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/rope_plugin.py]: Ok, no problem. Just curious :-)

File: spyderlib/utils/introspection/jedi_plugin.py (hunk 1 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/jedi_plugin.py]: Do you plan to remove or leave this `debug_print`? You haven't answered my question :-)

File: spyderlib/utils/introspection/jedi_plugin.py (hunk 2 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/jedi_plugin.py]: Ok, I just saw that you added a new (M) for modules.

@goanpeca, what do you think about it? Should we leave it or use a (P) one?

Comment (blink1073) [inline spyderlib/utils/introspection/jedi_plugin.py]: Because module != package, and P could also be property.

File: spyderlib/widgets/sourcecode/base.py [hunk 7]
```diff
@@ -61,15 +63,28 @@ def setup_appearance(self, size, font):
...
         if len(completion_list) == 1 and not automatic:
             self.textedit.insert_completion(completion_list[0])
             return
-        
+
         self.completion_list = completion_list
         self.clear()
-        self.addItems(completion_list)
+
+        icon_lut = {'instance': 'attribute.png',
```
Comment (ccordoba12) [inline spyderlib/widgets/sourcecode/base.py]: What does `lut` mean? I don't get it :-)

What about renaming this variable for `icons_map`?

**Comment by ccordoba12** [comment]
Other than my minor remarks, this is ready to merge!

Great work @blink1073!!

File: spyderlib/utils/introspection/jedi_plugin.py (hunk 2 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/jedi_plugin.py]: Ok, no problem then. Let's leave your new (M) :-)

File: spyderlib/widgets/sourcecode/base.py (hunk 7 above)
Comment (blink1073) [inline spyderlib/widgets/sourcecode/base.py]: Ha!  lut == look up table.  I'll change it.

File: spyderlib/utils/introspection/jedi_plugin.py (hunk 1 above)
Comment (blink1073) [inline spyderlib/utils/introspection/jedi_plugin.py]: Fixed now, with updated info!

**Comment by blink1073** [comment]
And I think I'm done.

**Comment by ccordoba12** [comment]
Let's wait for Travis before merging :-)
